### ISSUE TITLE
improve costMatrix compression

### ISSUE THREAD

**Issue by chobobobo** [ScreepsOCS/screeps.behaviour-action-pattern#1307]
from @Spedwards 
```
// TO
let serialized = bits.toString().replace(/((?:(\d)+,?)((?:\2+,?){2,}))/g, (m, p1, p2) => p2 + 'x' + m.replace(/,/g, '').length);

// FROM
let parsed = serialized.replace(/(\d+)x(\d+),?/g, (z, p1, p2) => _.repeat(p1 + ',', +p2)).split(',');
```

**Comment by Spedwards** [comment]
```javascript
const serialize = arr => arr
    .reduce((m, e, i) => {
        if (e !== arr[i - 1]) {
            m.push({count: 1, e});
        } else {
            m[m.length - 1].count++;
        }
        return m;
    }, [])
    .map(({count, e}) => count > 1 ? `${e}x${count}` : e)
    .join(',');
```

```javascript
const parse = serialize => {
    const arr = serialized.replace(/(\d+)x(\d+),?/g, (z, p1, p2) => _.repeat(p1 + ',', +p2)).split(',');
    arr.pop();
    return arr.map(s => +s);
};
```

**Comment by Spedwards** [comment]
Use the old method, the serialized string would be (for one of my rooms) 1487 characters. Using this method, I lowered it to 434.

**Comment by JRWR** [comment]
hrm, has anyone done any type of real compression like LZ4? could be used to save even more space

**Comment by chobobobo** [comment]
Yes, we tried that first, unfortunately when there data comes back from the segment some of the characters are wrong because of the way they are stored in the backend.

**Comment by chobobobo** [comment]
resolved by #1353
